src/common.rs

```diff
@@ -26,6 +26,103 @@ pub(crate) mod tests {
+
+       # Examples
+
+       The following example shows a single test case for the
+       DuplicatedKeyChecker:
+
+       ```
+       check_tester!{
+           DuplicatedKeyChecker;
+           duplicated_key_test => {
+               // First time seeing key "FOO" so is None
+               "FOO=BAR" => None,
+               // Second time seeing key "FOO" so is duplicate.
+               "FOO=BAR" => Some("The FOO key is duplicated"),
+           }
+       }
+       ```
+
+       The above example could be expanded as such:
+       ```
+        #[test]
+        fn duplicated_key_test() {
+            let mut checker = DuplicatedKeyChecker::default();
+
+            assert_eq!(None, checker.run(&line_entry(1, 2, "FOO=BAR")));
+            assert_eq!(Some(
+                Warning::new(
+                    line_entry(2, 2, "FOO=BAR")),
+                    "DuplicatedKey",
+                    "The FOO key is duplicated"
+                ),
+                checker.run(&line_entry(2, 2, "FOO=BAR"))
+            );
+        }
+       ```
+       As shown in the expanded example this macro will create the correct
+       Line_Entry for the checker and will also create the expected Warning
+       using the input and expected warning message.
+
+       ## Constant Warning Message
+
+       Note that the expected warning message does not need to be a string
+       literal; in the case where the warning message is always the same
+       a constant &str can be used:
+
+       ```
+        const WARNING: &str = "Invalid leading character detected";
+
+        check_tester!{
+            LeadingCharacterChecker;
+            leading_dot_is_not_valid => {
+                ".FOO=BAR" => Some(WARNING),
+            },
+            leading_number_is_not_valid => {
+                "1FOO=BAR" => Some(WARNING),
+            }
+        }
+       ```
+    */
+    #[cfg(test)]
+    #[macro_export]
+    macro_rules! check_tester {
+        (@token $t:tt $sub:expr) => {$sub};
+        (@count $($t:tt)*) => {<[()]>::len(&[$(check_tester!(@token $t ())),*])};
+        ($checker:ident;
+            $(
+                $test:ident => {$(
+                    $input:expr => $expected:expr,
+                )*}
+            ),* $(,)?) => {
+            $(
+                #[test]
+                fn $test() {
+                    let mut checker = $checker::default();
+                    let total = check_tester!(@count $($expected)*);
+                    let mut _line_number = 1;
+                    $(
+                        let line = line_entry(_line_number, total, $input);
+                        _line_number += 1;
+                        let result = checker.run(&line);
+                        let expected = ($expected).map(|e: &str| Warning::new(line, checker.name(), e));
+                        assert_eq!(expected, result);
+                    )*
+                }
+            )*
+        };
+    }
+
```

Comment:
In my opinion, declarative macros make a big contract here and are more difficult to understand and maintain in the future 🤔 
But I like going in this direction, I would prefer to use a generalized generic function, and reuse it in all tests in the same way.
Something like that:
```rust
    pub fn test_check<T>(checker: &mut T, asserts: Vec<(&str, Option<&str>)>) where T: Check {
        let mut line_number = 1;
        let total = asserts.len();

        for (input, expected) in asserts {
            let line = line_entry(line_number, total, input);
            line_number += 1;

            let result = checker.run(&line);
            let expected = expected.map(|e: &str| Warning::new(line, checker.name(), e));

            assert_eq!(result, expected);
        }
    }

```

As a result, the tests will look something like this:
```rust
    // UnorderedKeyChecker tests
    #[test]
    fn one_unordered_key_test() {
        test_check(&mut UnorderedKeyChecker::default(), vec![
            ("FOO=BAR", None),
            ("BAR=FOO", Some("The BAR key should go before the FOO key")),
        ]);
    }
    
    // QuoteCharacterChecker tests
    const WARNING: &str = "The value has quote characters (\', \")";

    #[test]
    fn with_single_quote_test() {
        test_check(&mut QuoteCharacterChecker::default(), vec![
            ("FOO=BAR", None),
            ("FOO='BAR'", Some(WARNING)),
            ("FOO='B\"AR'", Some(WARNING)),
            ("FOO=\'BAR BAR\'", None),
        ]);
    }
```

What do you think of this @mc1098  @dotenv-linter/core ?

Perhaps in this case a little more code, but it seems to me more clear and separated (no explicit contract with custom syntax)

Reply:
@DDtKey Thank you 😃!
I'm happy with either macro or function but I agree that the macro here makes it more complicated to change in the future. A function might be bit more code per test but is also more explicit, you can infer what that function is probably doing by it's signature but the same isn't quite true with the macro as it's bit more magic. 
I shall update this PR maybe today or tomorrow to use a generic function instead 🚀

Reply:
I have pretty much copied your suggestion. I changed the Vec for AsRef slice, which I think makes it more ergonomic and would avoid requiring allocation when called with an array over a Vec.  
https://github.com/dotenv-linter/dotenv-linter/blob/5fc4b217712428b9c4df9b1fe5878ae0189664c1/src/common.rs#L60-L63